---
title: "FONCTIOMIC_PICRUSt2_out"
auteur: "Dinh-Phuc Hoang"
date: "2025-02-27"
---

```{r}
library(readr)
library(dplyr)
library(tidyverse)
library("phyloseq")
library(ALDEx2)
library(DESeq2)
library(BiocParallel)
register(SnowParam(4))
```
Load necessary data: abundance data and metadata
```{r}
meta = read_delim("Data FONCTIOMIC/03_dimimos_terrain.csv",delim = ",",col_names = TRUE)
p2_EC = read_delim("Data FONCTIOMIC/EC_metagenome_out/pred_metagenome_unstrat.tsv.gz",delim = "\t",col_names = TRUE)
  p2_EC = as.data.frame(p2_EC)
  rownames(p2_EC) = p2_EC[,1]
  p2_EC = as.matrix(p2_EC[,-1])
  p2_EC = round(p2_EC)
p2_KO = read_delim("Data FONCTIOMIC/KO_metagenome_out/pred_metagenome_unstrat.tsv.gz",delim = "\t",col_names = TRUE)
    p2_KO = as.data.frame(p2_KO)
    rownames(p2_KO) = p2_KO[,1]
    p2_KO = as.matrix(p2_KO[,-1])
    p2_KO = round(p2_KO)
p2_PW = read_delim("Data FONCTIOMIC/pathways_out/path_abun_unstrat.tsv.gz",delim = "\t",col_names = TRUE)
    p2_PW = as.data.frame(p2_PW)
    rownames(p2_PW) = p2_PW[,1]
    p2_PW = as.matrix(p2_PW[,-1])
    p2_PW = round(p2_PW)
otu_table = read_delim("Data FONCTIOMIC/03_dimimos_terrain.seq_abundance.16S.tsv",delim = "\t",col_names = TRUE)
otu_table = as.data.frame(otu_table)
row.names(otu_table) = otu_table$sequence
otu_table = as.matrix(otu_table[,-1])
#tax_table = read_delim("",delim = "\t",col_names = FALSE)
phy_tree = read_tree("Data FONCTIOMIC/out.tre")
```
Load ref data: mapfile
```{r}
mapfile_EC = read_delim("description_mapfiles/ec_level4_info.tsv.gz",delim = "\t", col_names  = FALSE)
mapfile_KO = read_delim("description_mapfiles/ko_info.tsv.gz",delim = "\t", col_names = FALSE)
mapfile_PW = read_delim("description_mapfiles/metacyc_pathways_info.txt.gz",delim = "\t", col_names = FALSE)
colnames(mapfile_EC) = c("function","description")
colnames(mapfile_KO) = c("function","description")
colnames(mapfile_PW) = c("pathway","description")
```
1) Clean metadata (enlever les sample_id dèja enlevé dans les données)
#Define : p2_EC1 = tableau de l'abondance des EC pour les échantillons de contrôle
```{r}
meta = as.data.frame(meta)
rownames(meta) = meta$sample
meta$sample = as.factor(meta$sample)
meta$sample = factor(meta$sample, levels = c(sort(levels(meta$sample), decreasing = F)))
meta_cleaned = meta[colnames(p2_EC),]
meta_cleaned$groupe2 = paste(meta_cleaned$bloc,meta_cleaned$modality,sep = "_")
meta_cleaned = meta_cleaned %>% filter(campagne %in% c("0","1","4","6"))
meta_cleaned$campagne = as.factor(meta_cleaned$campagne)
meta_cleaned$campagne = relevel(meta_cleaned$campagne, ref = "0")
meta_cleaned = meta_cleaned %>% mutate(time = fct_recode(campagne, 
"0" = "0", "3" = "1", "51" = "4", "125" = "6")) %>% mutate(groups = fct_recode(groupe2, 
"Control Cropland" = "crops_control", "Amended Cropland" = "crops_residues", 
"Control Grassland" = "grasslands_control", "Amended Grassland" = "grasslands_residues"))
```
Phyloseq object
```{r}
ps = phyloseq(sample_data(meta),
                otu_table(as.matrix(otu_table), taxa_are_rows = FALSE),
                phy_tree(phy_tree))
```
Statistical analysis with ALDEx2 between 4 groups
```{r}
meta_cleaned$groupe2 = as.factor(meta_cleaned$groups)
meta_cleaned$groupe2 = relevel(meta_cleaned$groups, ref = "Control Cropland")
mm = model.matrix(~ meta_cleaned$groupe2 + meta_cleaned$time, data = meta_cleaned)
 x = aldex.clr (p2_EC, mm, mc.samples = 128, denom = "all")
glm.test = aldex.glm(x, mm)
glm.effect <- aldex.glm.effect(x)
p2_EC_aldex = glm.test[,13:20]
write.table(p2_EC_aldex, "ALDEx2/p2_EC_aldex.txt", sep = "\t", quote = FALSE, row.names = TRUE, col.names = TRUE)
```
PCA for EC (brut)
```{r}
p2_EC_wide = as.matrix(p2_EC)
p2_EC_wide = t(p2_EC_wide)
p2_EC_pca = vegan::rda(p2_EC_wide,scale=TRUE)
summary(p2_EC_pca)
```
PCA plot (EC)
```{r}
EC_pca_dat = as.data.frame(p2_EC_pca$CA$u[,1:2])
EC_pca_dat <- cbind(EC_pca_dat, meta_cleaned$groups)
colnames(EC_pca_dat) <- c("PC1","PC2","Groups")
ggplot(EC_pca_dat, aes(x = PC1, y = PC2, color = Groups, shape = Groups)) +
  ggtitle("PCA of EC abundance") +
  geom_point(alpha=1) +
  xlab("PC1: 33%")+
  ylab("PC2: 13%") +
  stat_ellipse(geom="polygon",type ="norm", level=0.9, alpha=0.1,
  aes(fill = Groups),
  color =NA) +
theme(plot.title = element_text(hjust = 0.5, size = 16, face = "bold"))
```
**bold**#Statistical analysis with DESSeq2 between 4 groups
Slice the data into groups by time

```{r}
p2_EC_cleaned = p2_EC[,rownames(meta_cleaned)]
dds_P2 = DESeqDataSetFromMatrix(countData =p2_EC_cleaned,
                                  colData = meta_cleaned,
                                  design = ~ time + groups, tidy = FALSE)
DAA_EC_P2 = DESeq(dds_P2, parallel = TRUE)
dds_P2_vst = vst(dds_P2,blind = FALSE)
res_EC = results(DAA_EC_P2, parallel = TRUE)
head(res_EC)
summary(res_EC)
```
PCA plotting
```{r}
pcaData = plotPCA(dds_P2, intgroup = c("time", "groups"), returnData = TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))
ggplot(pcaData, aes(x = PC1, y = PC2, color = groups, shape = time)) +
  geom_point(size =3) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
 ggtitle("PCA of EC abundance in output of PICRUSt2") +
 scale_color_manual(values = c("#FF6F91",
                                "#C34A36",
                                "#008F7A",
                                "#2C73D2"))+
  scale_shape_manual(values = c(16,15,17,18))+
  guides(color = guide_legend(title = "Groups"),shape = guide_legend(title = "Time")) +
  theme(plot.title = element_text(hjust = 0.5, size = 16, face = "bold"))
```

```{r}
dds_p2_vst = as.matrix(dds_MGS_vst[,-c(49)])
dds_MGS_vst = dds_MGS_vst[order(rownames(dds_MGS_vst)),
order(colnames(dds_MGS_vst))]
mat = dds_MGS_vst
mat = mat - rowMeans(mat)
dists_row = vegan::vegdist((dds_MGS_vst), method = "bray")
dists_col = vegan::vegdist(t(dds_MGS_vst), method = "bray")
```

```{r}
```